db/types/interval.py

```diff
@@ -0,0 +1,75 @@
+from sqlalchemy.dialects.postgresql import INTERVAL
+from sqlalchemy import func, case
+from sqlalchemy.types import TypeDecorator
+
+from db.types.exceptions import InvalidTypeParameters
+
+
+class Interval(TypeDecorator):
+    impl = INTERVAL
+    cache_ok = True
+
+    def __init__(self, *arg, **kwarg):
+        TypeDecorator.__init__(self, *arg, **kwarg)
+        self.validate_arguments()
+
+    def validate_arguments(self):
+        seconds_fields = {
+            'SECOND',
+            'DAY TO SECOND',
+            'HOUR TO SECOND',
+            'MINUTE TO SECOND',
+        }
+        other_fields = {
+            'YEAR',
+            'MONTH',
+            'DAY',
+            'HOUR',
+            'MINUTE',
+            'YEAR TO MONTH',
+            'DAY TO HOUR',
+            'DAY TO MINUTE',
+            'HOUR TO MINUTE',
+        }
+        all_fields = seconds_fields.union(other_fields)
+        if self.impl.precision is not None:
+            try:
+                assert isinstance(self.impl.precision, int)
+            except AssertionError:
+                raise InvalidTypeParameters('precision must be an integer')
+            try:
+                assert (
+                    self.impl.fields is None
+                    or self.impl.fields.upper() in seconds_fields
+                )
+            except AssertionError:
+                raise InvalidTypeParameters(
+                    'If precision and fields are both given,'
+                    ' seconds must be included in fields.'
+                )
+        elif self.impl.fields is not None:
+            try:
+                assert self.impl.fields.upper() in all_fields
+            except AssertionError:
+                raise InvalidTypeParameters(
+                    f'fields "{self.impl.fields}" is not in {all_fields}'
+                )
+
+    def column_expression(self, col):
```

Comment:
I don't have any issues with using `column_expression`,  I am just curious to know if you have thought about setting a local `intervalstyle` instead.

Reply:
The problem with that is we need to avoid `psycopg2` interpreting the result as an actual `INTERVAL`.  The column expression casts to `TEXT` at the DB layer, which then gets picked up by `psycopg2` as a python `str`.  The problem with letting `psycopg2` pick up `INTERVAL`s otherwise is that it uses a python `timedelta` to represent them, but `timedelta` makes some different choices (e.g., assuming 30 days can be accumulated into 1 month) and the conversion is lossy.  This would result in erroneous info being passed to the UI, and in fact make viewing some values (e.g., 37 days) impossible.

`intervalstyle` doesn't change the actual return type, and so doesn't really solve the problem by itself.  We'd still need to cast to text using a `column_expression`.  Given that, I opted to handle the formatting ourselves, since the PostgreSQL `iso_8601` style is slightly out-of-spec, and made a few choices that don't work that well for our use case (IMO).  It also makes the formatting that would happen completely visible, rather than hidden in the PostgreSQL docs.

Reply:
That makes sense, thanks for the explanation!